dev-docs/development_process.md

```diff
@@ -76,63 +76,141 @@ export DOCKER_BUILDKIT=1
 
+The first time a namespace is used, the Auth service in that namespace won't
+have any users. In order submit jobs to the namespace, make sure to include
+the `add_developers` step to the dev deploy. So a first dev deploy might look like:
 
-If the tests fail, you can then examine the Kubernetes logs for the service
-using something like
+```bash
+HAIL_DEFAULT_NAMESPACE=default hailctl dev deploy -b <github_user_name>/hail:<branch_name> -s deploy_batch,add_developers
+```
 
+After this dev deploy completes, you should be able to access your namespace
+by navigating to https://internal.hail.is/<username>/batch.
+
+To submit jobs to your dev namespace, you need to configure your local hail
+installation to point to the dev namespace. You can do this by running
+
+```bash
+hailctl dev config set default_namespace <my_namespace>
+hailctl auth login
 ```
+Then, any Hail Batch or Query-on-Batch script will submit to your dev namespace.
+
+You can also use the `HAIL_DEFAULT_NAMESPACE` environment variable to alter
+the namespace for individual terminal sessions or commands.
+
+NOTE: The dev deploy command above sets `HAIL_DEFAULT_NAMESPACE=default`. That's
+because `hailctl dev deploy` submits a request to the production CI service that
+lives in the `default` namespace. This does *not* mean that your branch will be deployed
+to the `default` namespace, it will always be deployed into your dev namespace.
+
+To examine the Kubernetes logs for the service in your namespace, use something like
+
+```bash
 kubectl -n <my_namespace> logs -l app=batch-driver --tail=999999 | less
 ```
 
-To check the MySQL database, you first need to find the name of the specific pod
-running an “admin-pod” and then log into that pod:
+To check the MySQL database in your namespace, you can log in to the database
+pod like so:
 
 ```
-kubectl -n <my_namespace> get pods -l app=admin-pod
-kubectl -n <my_namespace> exec -it <admin_pod_name> /bin/bash
+kubectl -n <my_namespace> exec -it db-0 /bin/bash
 $ mysql
 ```
 
+##### Alternatives to dev deploy
+
+There are three ways to deploy code into a dev namespace:
+1. dev deploy
+2. make deploy
+3. sync.py
+
+These are listed in order of broadest scope to narrowest scope, so it is never
+wrong to do something higher on the list, but it might be a slower feedback loop.
+
+A dev deploy is necessary under the following conditions:
+- A dev namespace has not been used before
+- You have added a migration that must be run against a service's database
+- Not all necessary services (like Auth) are deployed in the namespace and you
+  do not want to `make` deploy them individually.
+
+If you just want to redeploy a single service, you can do so with the following:
+
+```bash
+make -C batch deploy NAMESPACE=<my_namespace>
+```
+
+If you only want to make a Python code change to an already-running service
+in Kubernetes, you can run `devbin/sync.py` like the following:
+
+```bash
+python3 devbin/sync.py \
+    --namespace <my_namespace> \
```

Comment:
Can you make sure this works now with `batch-driver`? I was having problems with it because there's 2 containers in the batch-driver pod.

I also have `--path web_common/web_common /usr/local/lib/python3.9/dist-packages/` and `--path hail/python/hailtop /usr/local/lib/python3.9/dist-packages/`.

Reply:
Ah ya, added those paths too.

And ya I ran into the problem with `batch-driver` too but #13374 should have fixed it.